From 27280c0268d5c096d9eda61a8e5953ff0c04ce08 Mon Sep 17 00:00:00 2001
From: Jonathan Stroud <jonathan.stroud@amd.com>
Date: Wed, 18 Jun 2025 13:45:57 -0600
Subject: [PATCH] Reset the dwc3 USB core on init

If the dr_mode is changing from what was previously programmed, then a
core reset is required.

Signed-off-by: Jonathan Stroud <jonathan.stroud@amd.com>
---
 drivers/usb/dwc3/core.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 0c11bd1aebf7..b0dce0f9b604 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -1325,6 +1325,17 @@ int dwc3_core_init(struct dwc3 *dwc)
 	u32			reg;
 	int			ret;
 
+	// Reset the core on init
+	reg = dwc3_readl(dwc->regs, DWC3_GCTL);
+	if (DWC3_GCTL_PRTCAP(reg) != dwc->dr_mode) {
+		reg |= DWC3_GCTL_CORESOFTRESET;
+		reg &= ~DWC3_GCTL_PRTCAPDIR(DWC3_GCTL_PRTCAP_OTG);
+		dwc3_writel(dwc->regs, DWC3_GCTL, reg);
+		msleep(100);
+		reg &= ~DWC3_GCTL_CORESOFTRESET;
+		dwc3_writel(dwc->regs, DWC3_GCTL, reg);
+    }
+
 	hw_mode = DWC3_GHWPARAMS0_MODE(dwc->hwparams.hwparams0);
 
 	/*
-- 
2.34.1

